docs/rfc/block-registration.md

```diff
@@ -0,0 +1,371 @@
+
+
+## Previous attempts
+
+Initial support for server-defined block attributes was merged as part of [#2529](https://github.com/WordPress/gutenberg/pull/2529). PHP block type registrations are merged with those defined in the JavaScript runtime. While this enabled blocks to be defined within PHP, the majority of block types continue to be defined within JavaScript alone. The support was reserved for the exclusive use of [dynamic block types](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/block-tutorial/creating-dynamic-blocks/), in large part because <code>[edit and save behaviors](https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/block-edit-save/)</code> must still be implemented in JavaScript, and because a solution hadn't been considered for how to create individual block bundles during the build process, nor how to load such bundles efficiently if it were to come to be implemented.
+
+A demonstration for how block registration could be made filterable in PHP was explored in [#5802](https://github.com/WordPress/gutenberg/pull/5802). The purpose here was to explore how plugins could have better control over the registration.
+
+Another exploration in [#5652](https://github.com/WordPress/gutenberg/pull/5652) considered using JSON as a file format to share block type definitions between JavaScript and PHP. 
+
+### Conclusions
+
+* The current approaches to client-side block type registration cannot support the proposed requirement to have all block types known outside the browser context.
+* Using a statically-defined, JSON-formatted block type definition enables easy integration in both JavaScript and PHP runtimes.
+* Registering a block type in PHP would allow for attribute default values to be assigned as dynamically generated from some external state (e.g. a database value, or localized string).
+* By default, JSON does not support localization or dynamic values.
+* On the server, a block type `icon` property can only be assigned as a string and thus cannot support SVGs and component-based icons.
+
+---
+
+# Block Type Registration RFC
+
+## Introduction
+
+Blocks are the fundamental elements of the editor. They are the primary way in which plugins and themes can register their own functionality and extend the capabilities of the editor.
+
+## Registering a block type
+
+To register a new block type, start by creating a `block.json` file. This file:
+
+* Gives a name to the block type.
+* Defines some important metadata about the registered block type (title, category, icon, description, keywords).
+* Defines the attributes of the block type.
+* Links to the editor implementation of your block type, the save function and any other context-aware property.
+
+**Example:**
+
+
+```json
```

Comment:
We should either update these JSON snippets to omit the inline comments, so that they're in a valid format, or change this to be [JSON5](https://json5.org/), where the comments would be valid.

```suggestion
```json5
```

Reply:
Let's remove those comments as they are only added as hints in those examples.

Reply:
Done in 288853c.